components/raftstore-v2/src/operation/query/mod.rs

```diff
@@ -71,6 +74,151 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: raftstore::store::Transport>
         }
     }
 
+    #[inline]
+    pub fn on_refresh_region_buckets(
+        &mut self,
+        region_epoch: RegionEpoch,
+        mut buckets: Vec<raftstore::store::Bucket>,
+        bucket_ranges: Option<Vec<raftstore::store::BucketRange>>,
+    ) {
+        // bucket version layout
+        //   term       logical counter
+        // |-----------|-----------|
+        //  high bits     low bits
+        // term: given 10s election timeout, the 32 bit means 1362 year running time
+        let gen_bucket_version = |term, current_version| {
+            let current_version_term = current_version >> 32;
+            let bucket_version: u64 = if current_version_term == term {
+                current_version + 1
+            } else {
+                if term > u32::MAX.into() {
+                    error!(
+                        self.fsm.peer().logger,
+                        "unexpected term {} more than u32::MAX. Bucket
+                    version will be backward.",
+                        term
+                    );
+                }
+                term << 32
+            };
+            bucket_version
+        };
+
+        let region = self.fsm.peer().region();
+        if util::is_epoch_stale(&region_epoch, region.get_region_epoch()) {
+            error!(
+                 self.fsm.peer().logger,
+                "receive a stale refresh region bucket message";
+                "epoch" => ?region_epoch,
+                "current_epoch" => ?region.get_region_epoch(),
+            );
+
+            return;
+        }
+
+        let mut current_version = self
+            .fsm
+            .peer()
+            .region_buckets()
+            .as_ref()
+            .map(|b| b.meta.version)
+            .unwrap_or_default();
+        if current_version == 0 {
+            current_version = self
+                .fsm
+                .peer()
+                .last_region_buckets()
+                .as_ref()
+                .map(|b| b.meta.version)
+                .unwrap_or_default();
+        }
+        let mut region_buckets: BucketStat;
+        if let (Some(bucket_ranges), Some(peer_region_buckets)) =
+            (bucket_ranges, self.fsm.peer().region_buckets())
+        {
+            assert_eq!(buckets.len(), bucket_ranges.len());
+            let mut i = 0;
+            region_buckets = peer_region_buckets.clone();
+            let mut meta = (*region_buckets.meta).clone();
+            if !buckets.is_empty() {
+                meta.version = gen_bucket_version(self.fsm.peer().term(), current_version);
+            }
+            meta.region_epoch = region_epoch;
+            for (bucket, bucket_range) in buckets.into_iter().zip(bucket_ranges) {
```

Comment:
This code is too hard to read. I suggest (as a first step) adding some comments about what is `i` and what it's used for.

Reply:
fixed